xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+          rdsResponse.getNonce(), "Broken RDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken RDS response"));
+      return;
+    }
+
+    // Validate and cache information from each RouteConfiguration message.
+    Map<String, String> clusterNames = new HashMap<>();
+    for (RouteConfiguration routeConfig : routeConfigs) {
+      String clusterName = processRouteConfig(routeConfig);
+      if (clusterName == null) {
+        adsStream.sendNackRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+            rdsResponse.getNonce(),
+            "Received RouteConfigurations without cluster information for requested hostname");
+        return;
+      }
+      clusterNames.put(routeConfig.getName(), clusterName);
+    }
+    routeConfigNamesToClusterNames.putAll(clusterNames);
+
+    adsStream.sendAckRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+        rdsResponse.getVersionInfo(), rdsResponse.getNonce());
+
+    // Notify the ConfigWatcher if this RDS response contains the most recently requested
+    // RDS resource.
+    if (clusterNames.containsKey(adsStream.rdsResourceName)) {
+      ConfigUpdate configUpdate =
+          ConfigUpdate.newBuilder()
+              .setClusterName(clusterNames.get(adsStream.rdsResourceName))
+              .build();
+      configWatcher.onConfigChanged(configUpdate);
+    }
+    // Do not notify an error to the ConfigWatcher. RDS protocol is incremental, not receiving
+    // requested RouteConfiguration in this response does not imply absence.
+  }
+
+  /**
+   * Processes RouteConfiguration message (from an resource information in an LDS or RDS
+   * response), which may contain a VirtualHost with domains matching the "xds:"
+   * URI hostname directly in-line. Returns the clusterName found in the VirtualHost
+   * message with domains matching the "xds:" URI hostname. Returns {@code null} if such a
+   * clusterName cannot be resolved.
+   *
+   * <p>Note we only validate VirtualHosts with domains matching the "xds:" URI hostname as we
+   * never care about others.
+   */
+  @Nullable
+  private String processRouteConfig(RouteConfiguration config) {
+    List<VirtualHost> virtualHosts = config.getVirtualHostsList();
+    String clusterName = null;
+    // Proceed with the virtual host that has longest wildcard matched domain name with the
+    // hostname in original "xds:" URI.
+    int matchingLen = -1;  // longest length of wildcard pattern that matches host name
+    for (VirtualHost vHost : virtualHosts) {
+      for (String domain : vHost.getDomainsList()) {
+        // We only validate data for VirtualHosts that matches hostname.
+        if (matchHostName(hostName, domain) && domain.length() > matchingLen) {
```

Comment:
we don't want 0-length domain matching and succeeding right? Of course this is a edge case and will never happen but shouldn't `matchingLen` be initialized to `0`?

Reply:
IMO, `matchHostName(...)` should not return `true` for any of it's argument being empty string. Otherwise, it doesn't make sense and is counter-intuitive. Given that, initializing `matchingLen` to `-1` or `0` does the same thing. I'd prefer `-1` as the initial default for a flag value.